Add Check Support for Java 21 Record Pattern : IllegalType

child of #14890 
I have read check documentation: https://checkstyle.org/checks/coding/illegaltype.html#IllegalType
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
PS D:\CS\test> javac src/RecordPatterns.java
PS D:\CS\test> cat src/RecordPatterns.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

public class RecordPatterns {
    record Box<T>(T t) { }

    public void TestIllegalType() {
        List<LinkedHashMap<Integer, Integer>> l // violation
                = new ArrayList<>();

        if (l instanceof List<LinkedHashMap<Integer, Integer>> linkedHashMap) { // violation
            System.out.println(linkedHashMap);
        }
         LinkedHashMap<Integer, Integer> l2 // violation
                = new LinkedHashMap<>();
        Box<LinkedHashMap<Integer,Integer>> box = new Box<>(l2);  // violation

        if (box instanceof Box<LinkedHashMap<Integer,Integer>>(var linkedHashMap)) { // expected violation
            System.out.println(linkedHashMap);
        }
    }
}
PS D:\CS\test> cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="IllegalType"/>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:9:14: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:12:31: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:15:10: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:17:13: Usage of type 'LinkedHashMap' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 4 errors.
```
---

**Describe what you expect in detail.**

I expect a violation on the type parameters of the record pattern.  The type parameter of the record pattern is of illegal type. 

---

> The is slightly confusing as the line says missing a violation but has a comment which says violation. I had to make sure I was finding the right line 12.

Yes sorry my wording could be confusing. I wanted to say that I want a violation on record pattern (line 19)  like we have a violation on the type pattern (line 12). 

> Your example only used var which is not a strict type, but we will be expanding this check for things like (random example from JEP):

yes ofc this should be solved within this issue. I just was trying to make the example with the default config. here is extra example:

```Java
PS D:\CS\test> javac src/RecordPatterns.java                                               
PS D:\CS\test> cat src/RecordPatterns.java                                                 
record ColoredPoint(Point p, String c) { } // violation
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }  // 2 violations
record Point(int x, int y) {}

public class RecordPatterns {
    public void TestIllegalType(Object obj) {
        if (obj instanceof Rectangle r) { }  // violation
        if (obj instanceof ColoredPoint l)  // violation
        if (obj instanceof Rectangle(ColoredPoint x, ColoredPoint y)) { } // 2 violations, expected  3
        if (obj instanceof Rectangle(ColoredPoint(Point(int x1,int x2),_),_)) { } // no violations, expected 3
    }
}
PS D:\CS\test> cat config.xml                                                              
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
 <module name="TreeWalker">
    <module name="IllegalType">
        <property name="illegalClassNames" value="Rectangle, ColoredPoint, Point"/>
    </module>
 </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.14.2-all.jar -c config.xml  src/RecordPatterns.java
Starting audit...
[ERROR] D:\CS\test\src\RecordPatterns.java:1:21: Usage of type 'Point' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:2:18: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:2:42: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:7:28: Usage of type 'Rectangle' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:8:28: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:9:38: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
[ERROR] D:\CS\test\src\RecordPatterns.java:9:54: Usage of type 'ColoredPoint' is not allowed. [IllegalType]
Audit done.
Checkstyle ends with 7 errors.

```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheck.java
@@ -143,7 +143,9 @@ import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
  * RECORD_DEF</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_COMPONENT_DEF">
- * RECORD_COMPONENT_DEF</a>.
+ * RECORD_COMPONENT_DEF</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_PATTERN_DEF">
+ * RECORD_PATTERN_DEF</a>.
  * </li>
  * </ul>
  * <p>
@@ -264,6 +266,7 @@ public final class IllegalTypeCheck extends AbstractCheck {
             TokenTypes.PATTERN_VARIABLE_DEF,
             TokenTypes.RECORD_DEF,
             TokenTypes.RECORD_COMPONENT_DEF,
+            TokenTypes.RECORD_PATTERN_DEF,
         };
     }
 
@@ -299,6 +302,7 @@ public final class IllegalTypeCheck extends AbstractCheck {
                 visitVariableDef(ast);
                 break;
             case TokenTypes.RECORD_COMPONENT_DEF:
+            case TokenTypes.RECORD_PATTERN_DEF:
                 checkClassName(ast);
                 break;
             case TokenTypes.PARAMETER_DEF:

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/IllegalTypeCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/coding/IllegalTypeCheck.xml
@@ -83,7 +83,7 @@
                       type="boolean">
                <description>Control whether to validate abstract class names.</description>
             </property>
-            <property default-value="ANNOTATION_FIELD_DEF,CLASS_DEF,INTERFACE_DEF,METHOD_CALL,METHOD_DEF,METHOD_REF,PARAMETER_DEF,VARIABLE_DEF,PATTERN_VARIABLE_DEF,RECORD_DEF,RECORD_COMPONENT_DEF"
+            <property default-value="ANNOTATION_FIELD_DEF,CLASS_DEF,INTERFACE_DEF,METHOD_CALL,METHOD_DEF,METHOD_REF,PARAMETER_DEF,VARIABLE_DEF,PATTERN_VARIABLE_DEF,RECORD_DEF,RECORD_COMPONENT_DEF,RECORD_PATTERN_DEF"
                       name="tokens"
                       type="java.lang.String[]"
                       validation-type="tokenSet">

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/IllegalTypeCheckTest.java
@@ -456,4 +456,34 @@ public class IllegalTypeCheckTest extends AbstractModuleTestSupport {
                 getPath("InputIllegalTypeAbstractClassNameFormat.java"),
                 expected);
     }
+
+    @Test
+    public void testIllegalTypeWithRecordPattern() throws Exception {
+        final String[] expected = {
+            "28:25: " + getCheckMessage(MSG_KEY, "Point"),
+            "29:22: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+            "29:46: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+            "39:28: " + getCheckMessage(MSG_KEY, "Rectangle"),
+            "40:28: " + getCheckMessage(MSG_KEY, "Rectangle"),
+            "40:38: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+            "40:54: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+            "45:28: " + getCheckMessage(MSG_KEY, "Rectangle"),
+            "45:38: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+            "45:51: " + getCheckMessage(MSG_KEY, "Point"),
+            "50:28: " + getCheckMessage(MSG_KEY, "Point"),
+            "58:9: " + getCheckMessage(MSG_KEY, "LinkedHashMap"),
+            "61:9: " + getCheckMessage(MSG_KEY, "Box"),
+            "61:13: " + getCheckMessage(MSG_KEY, "LinkedHashMap"),
+            "66:28: " + getCheckMessage(MSG_KEY, "Box"),
+            "66:32: " + getCheckMessage(MSG_KEY, "LinkedHashMap"),
+            "76:18: " + getCheckMessage(MSG_KEY, "Point"),
+            "77:18: " + getCheckMessage(MSG_KEY, "Rectangle"),
+            "77:28: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+            "77:41: " + getCheckMessage(MSG_KEY, "Point"),
+            "83:18: " + getCheckMessage(MSG_KEY, "ColoredPoint"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputIllegalTypeWithRecordPattern.java"),
+                expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWithRecordPattern.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/illegaltype/InputIllegalTypeWithRecordPattern.java
@@ -1,0 +1,87 @@
+/*
+IllegalType
+validateAbstractClassNames = (default)false
+illegalClassNames = HashMap, HashSet, LinkedHashMap, LinkedHashSet, TreeMap, TreeSet, \
+                    java.util.HashMap, java.util.HashSet, java.util.LinkedHashMap, \
+                    java.util.LinkedHashSet, java.util.TreeMap, java.util.TreeSet, \
+                    Rectangle, ColoredPoint, Point, Box
+legalAbstractClassNames = (default)
+ignoredMethodNames = (default)getEnvironment, getInitialContext
+illegalAbstractClassNameFormat = (default)^(.*[.])?Abstract.*$
+memberModifiers = (default)
+tokens = (default)ANNOTATION_FIELD_DEF, CLASS_DEF, INTERFACE_DEF, METHOD_CALL, METHOD_DEF, \
+         METHOD_REF, PARAMETER_DEF, VARIABLE_DEF, PATTERN_VARIABLE_DEF, RECORD_DEF, \
+         RECORD_COMPONENT_DEF, RECORD_PATTERN_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.illegaltype;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+public class InputIllegalTypeWithRecordPattern {
+
+    record ColoredPoint(Point p, String c) { } // violation, 'Usage of type 'Point' is not allowed.'
+    record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
+    // 2 violations above:
+    //                    'Usage of type 'ColoredPoint' is not allowed.'
+    //                    'Usage of type 'ColoredPoint' is not allowed.'
+    record Point(int x, int y) { }
+    record Box<T>(T t) { }
+
+
+    public void testWithInstanceOf(Object obj) {
+        // violation below, 'Usage of type 'Rectangle' is not allowed.'
+        if (obj instanceof Rectangle r) { }
+        if (obj instanceof Rectangle(ColoredPoint x, ColoredPoint y)) { }
+        // 3 violations above:
+        //                  'Usage of type 'Rectangle' is not allowed.'
+        //                  'Usage of type 'ColoredPoint' is not allowed.'
+        //                  'Usage of type 'ColoredPoint' is not allowed.'
+        if (obj instanceof Rectangle(ColoredPoint(Point(int _,int _),_),_)) { }
+        // 3 violations above:
+        //                  'Usage of type 'Rectangle' is not allowed.'
+        //                  'Usage of type 'ColoredPoint' is not allowed.'
+        //                  'Usage of type 'Point' is not allowed.'
+        if (obj instanceof Point(_,_)) { }
+        // violation above, 'Usage of type 'Point' is not allowed.'
+
+    }
+
+    public void testWithInstanceOfWithGenerics() {
+
+        // violation below, 'Usage of type 'LinkedHashMap' is not allowed.'
+        LinkedHashMap<Integer, Integer> l2
+                = new LinkedHashMap<>();
+
+        Box<LinkedHashMap<Integer,Integer>> box = new Box<>(l2);
+        // 2 violations above:
+        //                  'Usage of type 'Box' is not allowed.'
+        //                  'Usage of type 'LinkedHashMap' is not allowed.'
+
+        if (box instanceof Box<LinkedHashMap<Integer,Integer>>(var linkedHashMap)) {}
+        // 2 violations above:
+        //                  'Usage of type 'Box' is not allowed.'
+        //                  'Usage of type 'LinkedHashMap' is not allowed.'
+
+    }
+
+    public void testWithSwitch(Object obj) {
+        switch (obj) {
+            // violation below,  'Usage of type 'Point' is not allowed.'
+            case Point(_,_) -> System.out.println("point");
+            case Rectangle(ColoredPoint(Point(_, _),_),_) -> System.out.println("rectangle");
+              // 3 violations above:
+              //                  'Usage of type 'Rectangle' is not allowed.'
+             //                  'Usage of type 'ColoredPoint' is not allowed.'
+            //                    'Usage of type 'Point' is not allowed.'
+            // violation below,  'Usage of type 'ColoredPoint' is not allowed.'
+            case ColoredPoint _ -> System.out.println("coloredPoint");
+            default -> throw new IllegalStateException("Unexpected value: " + obj);
+        }
+    }
+}

--- a/src/xdocs/checks/coding/illegaltype.xml
+++ b/src/xdocs/checks/coding/illegaltype.xml
@@ -101,6 +101,8 @@
                     RECORD_DEF</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_COMPONENT_DEF">
                     RECORD_COMPONENT_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_PATTERN_DEF">
+                    RECORD_PATTERN_DEF</a>
                   .
               </td>
               <td>
@@ -126,6 +128,8 @@
                     RECORD_DEF</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_COMPONENT_DEF">
                     RECORD_COMPONENT_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_PATTERN_DEF">
+                    RECORD_PATTERN_DEF</a>
                   .
               </td>
               <td>3.2</td>

